### ISSUE TITLE
Remove `substrate-test-runtime`

### ISSUE THREAD

**Issue by bkchr** [paritytech/substrate#9782]
The `substrate-test-runtime` is some special kind of runtime that incorporates some FRAME stuff and some custom logic. We need to check all the usages of `substrate-test-runtime`, but this should all be solvable with a pure FRAME based runtime. It will probably require some special pallets as part of the runtime, but that should be no problem. We also will require more than one runtime for testing. One runtime for Babe and one for Aura (and maybe more in the future). So, that we have one runtime per authoring algorithm that can be used to tests them properly.

As an example [Cumulus test environment](https://github.com/paritytech/cumulus/tree/master/test) can be used. The runtime/service/client there are all ready to be used in individual tests or in integration tests. Something similar should be created for Substrate.

**Comment by gilescope** [comment]
Sorry I should ask questions sooner rather than waste time trying to figure things out by myself. The substrate-test-runtime is used by the cumulus test environment so I am assuming that we are looking to make the substrate-test-runtime more general and split it into a babe and a aura (and maybe a manual seal one?).

I have been copying across the cumulus as an example, stripping out the relay chain and seeing if I can make that the basis to start from and port the tests to use that rather than using the substrate-test-runtime directly. (and then subsume the bits of substrate-test-runtime that are still needed).

So there will be a substrate/tests dir like cumulus and polkadot and the substrate/test-utils dir will hopefully disappear.

I am not sure if there was a question in there - more a statement of what I am trying to do...

**Comment by bkchr** [comment]
> The substrate-test-runtime is used by the cumulus test environment

No it is not used in Cumulus.



> the substrate-test-runtime more general and split it into a babe and a aura (and maybe a manual seal one?)

Yes that should be the way forward. A manual seal one should probably used in most of the tests that just want to interact with "a runtime" to test stuff. And a Babe + Aura version should be used by the respective crates to test their code, these runtimes could be stripped down to the minimum. 



> So there will be a substrate/tests dir like cumulus and polkadot and the substrate/test-utils dir will hopefully disappear.

This is only about Substrate, nothing that touches Polkadot. `test-utils` also contains the `test-client` which would stay, but could also be moved `tests`.

**Comment by gilescope** [comment]
Ah yes I was thinking of the references to `sustrate-test-client` - there are no direct references to the `substrate-test-runtime`.

**Comment by bkchr** [comment]
As I just the days thought about all this code here:
https://github.com/paritytech/substrate/blob/a76ed8c6d4c0655f208ff51684efbeb64d32dea7/test-utils/runtime/src/lib.rs#L696-L1183

In this new "model" we should put this code in its own crate that is compiled to WASM etc and is only used by `sp-api-test`.

**Comment by gilescope** [comment]
Agreed. impl_runtime_apis! feels like a hack to do hand rolled stuff.

**Comment by bkchr** [comment]
> Agreed. impl_runtime_apis! feels like a hack to do hand rolled stuff.

Not sure what you mean by hand rolled stuff here. You can just use the macro, also for `sp-api-test`
